docs/DesignDocs/ReproducibleBuildUsingNix.md

```diff
@@ -0,0 +1,397 @@
+
+### The Nix Repo
+
+To build, a clone of the nix packages repo must be present. There is a default clone installed when nix 
+is installed. 
+
+The nix packages repo serves two purposes, somewhat at odds with each other.
+
+- The package repository for a linux distro "NixOS" which will eventully be a standalone distro. In that 
+  case packages are chosen for release branches with an eye to mutual compatibility and working well within
+  that defined universe.  
+
+- The package repository is also used for producing packages for other linux distros (ubuntu, debian,
+  redhat, fedora, arch, gentoo) with easily auditable dependencies. In this case, it is necessary to
+  consider specific compatible versions of packages.  In particular, glibc must be an equal or older
+  revision than the glibc of the platform or the program will not dynamically link.  This is by design.
+  Ubuntu 18.04 is limited to glibc 2.27. The current lts release 20.04 usues glibc 2.31 and fedora 33 
+  uses libc-2.32.  The nixpkgs branch release-20.09 is based on glibc 2.32.  The somewhat older branch
+  20.03 uses libc-2.30 so can work on the current common distros, including the current rhel.  
+
+  While it is not difficult to change the version of glibc in a custom repo branch, since glibc is linked 
+  by every package, to do so would require a custom repo build. In that case, it would require a custom 
+  cache to provide prebuilding, or everything would require everything be built from source..
+
+### Caching and Hydra
+
+The official branches in nixpkgs are prebuilt and available at cache.nixos.org. Caching is keyed to the build outputs
+sha digest, so custom environments will be able to use those prebuilts so long as no changes have been made. 
+If any changes are made to that package, any package related to the changed package will need to be custom
+built and cached in nix store.
+
+
+## A Nix build system for OpenEnclave SDK
+
+We propose to add an optional nix based build environment for OpenEnclave SDK using docker containers.
+
+This would consist of :
+- Dockerfile.nix and build-oe-nix-build.sh for building the oe nix docker image locally.
+- build_openenclave.sh for building the oe deb package via a nix environment.
+- Dockerfile.test and test-oe-nix-build.sh for verifying that package is functional.
+
+The process builds a reproducible .deb package of openenclave-sdk. 
+
+This process can be integrated with ADO or CPDX CI, but that is not part of this submission. We do not currently
+recommend depending exclusively on a nix environment for production packaging and CI.
+
+### Dockerfile.nix
+
+The dockerfile is passed up to four build args:
+
+ - BASE_IMAGE allows for different os base images, in particular it allows the substitution of a
+   CPDX base image rather than the generic ubuntu 20.04. As was previously mentioned, we are basing on 
+   ubuntu 20.04 because package versions align best with nixpkgs release-20.03.
```

Comment:
Does this mean the OE SDK binaries produced might not work on Ubuntu 18.04 as the glibc version dynamically linked to the produced OE SDK binaries is v2.32, higher than the v2.27 available on Ubuntu 18.04?

Reply:
nixpkgs release-20.03 uses glibc v2.30, which is compatible with ubuntu but not ubuntu 18.04 which uses 2.27.

I tried overriding glibc version and was successful, but it meant removing all the patches the nixpkgs expressison contained, and eliminating all caching since glibc is a dependency for every package in the repo.  19.09 did contain an old enough glibc for ubuntu 18.04 but it is end of life and no longer supported. I was warned away from using it. Also, it contained a number of packages (openssl etc) that were very old.  I don't quite know why the major gap, but there it is.

This supports my impression that we need to manage our own fork of nixpkgs if we are to use this in production.

Reply:
Quick background to clarify.  glibc is designed no to dynalink with platforms that have an older version of glibc than the client application. So a client that has been linked with glibc 2.27 will work transparently on a platform that has glibc 2.30, but a client that is linked with 2.30 will not work on a platform that has glibc 2.27.  It is not practical to upgrade the glibc for the platform.

The nixpkgs repo is  designed to implement a quasi-distro, nixos, which is based on the nix system.  The packages in the 
nixpkgs repo are intended to serve that purpose.  It is not a non goal that nixpkgs supports building for other distros, but the priority is ambiguous.  The straightforward fix is to build a fork of nixpkgs which contains the expressions for a set of vetted packages, but that would be a large job to do right, and an ongoing commitment of resources.

Reply:
https://github.com/intel/linux-sgx/blob/544159a94a59547a18b7bff1d28a94bb14053572/linux/reproducibility/Dockerfile refers to a "/nix/store/681354n3k44r8z90m35hm8945vsp95h1-glibc-2.27". Is the problem with directly using the glibc-2.27 package that no pairing between the package and a .drv file can be established?

Reply:
Correct, plus that derivation is coming from (I beleive) 19.09 or even 19.03, both EOL and deprecated.

Reply:
IMO we should set up a common fork for nixpkgs that can be used by all of our partners and ourselves, which can be curated, and package versions set to reflect the OS versions we wish to support. Then the issue of packager version overrides etc disappears along with any possible reservations about trusting the packages